master/internal/logpattern/logpattern.go

```diff
@@ -0,0 +1,226 @@
+package logpattern
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/task/tasklogger"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
+
+	"github.com/uptrace/bun"
+)
+
+const regexCacheSize = 256
+
+var defaultSingleton *logPatternPolicies
+
+type logPatternPolicies struct {
+	regexCache *lru.Cache[string, *regexp.Regexp]
+}
+
+// New create the log pattern policies singleton.
+func New(ctx context.Context) (*logPatternPolicies, error) { //nolint: revive
+	regexCache, err := lru.New[string, *regexp.Regexp](regexCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("creating LRU cache: %w", err)
+	}
+
+	return &logPatternPolicies{
+		regexCache: regexCache,
+	}, nil
+}
+
+func (l *logPatternPolicies) getCompiledRegex(regex string) (*regexp.Regexp, error) {
+	if compiledRegex, ok := l.regexCache.Get(regex); ok {
+		return compiledRegex, nil
+	}
+
+	compiledRegex, err := regexp.Compile(regex)
+	if err != nil {
+		return nil, fmt.Errorf("compiling regex '%s': %w", regex, err)
+	}
+	l.regexCache.Add(regex, compiledRegex)
+
+	return compiledRegex, nil
+}
+
+func (l *logPatternPolicies) monitor(ctx context.Context,
+	taskID model.TaskID, logs []*model.TaskLog, policies expconf.LogPoliciesConfig,
+) error {
+	if len(policies) == 0 {
+		return nil
+	}
+
+	// TODO when we add rm specific log grabbing we will need to also monitor them.
+	for _, log := range logs {
+		if log.AgentID == nil {
+			return fmt.Errorf("agentID must be non nil to monitor logs")
+		}
+
+		for _, policy := range policies {
+			compiledRegex, err := l.getCompiledRegex(policy.Pattern())
+			if err != nil {
+				return err
+			}
+
+			if compiledRegex.MatchString(log.Log) {
+				switch policy.Action().GetUnionMember().(type) {
+				case expconf.LogActionCancelRetries:
+					if err := addDontRetry(
+						ctx, model.TaskID(log.TaskID), *log.AgentID, policy.Pattern(), log.Log,
+					); err != nil {
+						return fmt.Errorf("adding don't retry: %w", err)
+					}
+
+				case expconf.LogActionExcludeNode:
+					if err := addRetryOnDifferentNode(
+						ctx, model.TaskID(log.TaskID), *log.AgentID, policy.Pattern(), log.Log,
+					); err != nil {
+						return fmt.Errorf("adding retry on different node: %w", err)
+					}
```

Comment:
re: webhooks pr

can the webhook events get written here?

Reply:
talked offline and decided to make webhooks scan their logs